src/EndpointRegistry.sol

```diff
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: Apache 2
+pragma solidity >=0.6.12 <0.9.0;
+
+/// @dev This contract is responsible for handling the registration of Endpoints.
+abstract contract EndpointRegistry {
+    constructor() {
+        _checkEndpointsInvariants();
+    }
+
+    /// @dev Information about registered endpoints.
+    struct EndpointInfo {
+        // whether this endpoint is registered
+        bool registered;
+        // whether this endpoint is enabled
+        bool enabled;
+        uint8 index;
+    }
+
+    /// @dev Bitmap encoding the enabled endpoints.
+    /// invariant: forall (i: uint8), enabledEndpointBitmap & i == 1 <=> endpointInfos[i].enabled
+    struct _EnabledEndpointBitmap {
+        uint64 bitmap;
+    }
+
+    /// @dev Total number of registered endpoints. This number can only increase.
+    /// invariant: numRegisteredEndpoints <= MAX_ENDPOINTS
+    /// invariant: forall (i: uint8),
+    ///   i < numRegisteredEndpoints <=> exists (a: address), endpointInfos[a].index == i
+    struct _NumRegisteredEndpoints {
+        uint8 num;
+    }
+
+    // TODO: should we just increase this to 255?
+    uint8 constant MAX_ENDPOINTS = 64;
```

Comment:
At first glance I don't see a problem increasing to 255? But tbh even 64 endpoints sounds kind of crazy so I'm happy to leave this as is

Reply:
will revisit this later